### PR TITLE
Add telemetry & rate limiter interceptor

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Add telemetry RPC interceptor
* Add rate limiter RPC interceptor
* Refactor matching service removing unnecessary API metrics
* Refactor history service removing unnecessary API metrics

**Why?**
Unified API metrics / rate limit

**How did you test it?**
CICD

**Potential risks**
N/A

### CODE REVIEW

**Comment by wxing1292** [comment]
this PR targets 1.8.x

File: common/rpc/interceptor/telemetry.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+func (i *TelemetryInterceptor) Intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	_, methodName := SplitMethodName(info.FullMethod)
+	scope, ok := i.scopes[methodName]
+	if !ok {
+		i.logger.Error(fmt.Sprintf("unable to find scope for %v", methodName))
```
Comment (alexshtin) [inline common/rpc/interceptor/telemetry.go:72]: `fmt.Sprintf` shouldn't be used with logger. You can use `tag.Value` if you don't want to create custom tag.

Comment (alexshtin) [reply common/rpc/interceptor/telemetry.go:72]: This is probably not a big deal for `Error` messages as they are always on, but I cleaned up entire project some time ago and I would like to keep this state.

File: common/rpc/interceptor/telemetry.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	}
+	i.metricsClient.IncCounter(scope, metrics.ServiceRequests)
+	timer := i.metricsClient.StartTimer(scope, metrics.ServiceLatency)
+	defer timer.Stop()
+
+	resp, err := handler(ctx, req)
+	if err != nil {
+		i.handleError(scope, err)
+		return nil, err
```
Comment (alexshtin) [inline common/rpc/interceptor/telemetry.go:81]: I would just pass `resp` here:
```suggestion
		return resp, err
```

File: service/history/handler.go [hunk 3]
```diff
@@ -256,113 +250,78 @@ func (h *Handler) Watch(*healthpb.HealthCheckRequest, healthpb.Health_WatchServe
...
-	if ok := h.rateLimiter.Allow(); !ok {
-		return nil, h.error(errHistoryHostThrottle, scope, namespaceID, "")
+		return nil, h.convertError(errNamespaceNotSet)
 	}
 
 	heartbeatRequest := request.HeartbeatRequest
 	taskToken, err0 := h.tokenSerializer.Deserialize(heartbeatRequest.TaskToken)
 	if err0 != nil {
-		return nil, h.error(errDeserializeTaskToken.MessageArgs(err0), scope, namespaceID, "")
+		return nil, h.convertError(errDeserializeTaskToken.MessageArgs(err0))
```
Comment (alexshtin) [inline service/history/handler.go:264]: `convertError` is another good candidate for interceptor.

**Review by alexshtin — APPROVED** [review]
This is great change! Why you did not do the same for `workflowHandler.go`?

File: service/history/handler.go (hunk 3 above)
Comment (wxing1292) [reply service/history/handler.go:264]: all common API handler logic goes into interceptor

File: common/rpc/interceptor/telemetry.go (hunk 2 above)
Comment (wxing1292) [reply common/rpc/interceptor/telemetry.go:79]: it is preferable to make code explicit, below is easier to read and debug
```go
if err != nil {
	// error handling logic
	return nil, err
}
return resp, nil
```

File: common/rpc/interceptor/api_name.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import "strings"
+
+func SplitMethodName(
+	fullMethodName string,
+) (string, string) {
+	fullMethodName = strings.TrimPrefix(fullMethodName, "/") // remove leading slash
+	if i := strings.Index(fullMethodName, "/"); i >= 0 {
+		return fullMethodName[:i], fullMethodName[i+1:]
+	}
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/rpc/interceptor/api_name.go:35]: the impl is direct copy paste from here: https://github.com/grpc-ecosystem/go-grpc-prometheus/blob/9abf3eb82b4a313b1a28f370a7ef8ffe0667c681/util.go#L31

**Comment by wxing1292** [comment]
> This is great change! Why you did not do the same for `workflowHandler.go`?

that is part 2: #1285

File: common/rpc/interceptor/api_name.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/rpc/interceptor/api_name.go:35]: i am going to land this PR first, this will be followed by #1285 #1286

**Pull request merged**